### ISSUE TITLE
Propagate all BasicDeliverEventArgs arguments 

### ISSUE THREAD

**Issue by erezbartal** [Particular/NServiceBus.RabbitMQ#103]
When sending a message from native RabbitMQ API to a NSB endpoint not all BasicDeliverEventArgs are being propogated into the bus.CurrentMessageContext object.

These object could be Exchange, RoutingKey, UserID, etc....

Why do I need these parameters: In an IoT environment where non .NET devices connects directly to the broker there is a need to know which device sends the message. The device code can add a property to the message identifying who it is but this is not secure because a compromised code can impersonate to another device. We need to find a way that the broker infrastructure will identify the client.

One of the options to do it is to use client side certificate or any other rabbit routing mechanism. All these solutions eventually mapped to one or more parameters in the BasicDeliverEventArgs or in BasicDeliverEventArgs.BasicProperties. Unfortunately these parameters are not copied into the TransportMessage object (and eventually into the bus.CurrentMessageContext object) at MessageConverter.ToTransportMessage method

**Comment by andreasohlund** [comment]
:+1:

**Comment by SimonCropp** [comment]
@andreasohlund any objections to simply adding them as headers?

**Comment by bording** [comment]
@andreasohlund Is this something that should be added to the 4.0 milestone?

**Comment by andreasohlund** [comment]
I'd say no, we can add this in a minor

**Comment by andreasohlund** [comment]
One option to solve this would be to add the `B` to the message context passed to the core pipeline

**Comment by bording** [comment]
> One option to solve this would be to add the `BasicDeliverEventArgs` to the message context passed to the core pipeline

> If we go that route we should also consider exposing it on the error context

How are you thinking we could do that?

**Comment by andreasohlund** [comment]
Just add it to the contextbag?

**Comment by bording** [comment]
That's not really a great experience to get it back out, but maybe we could add an extension method or something to help.

However, there is no `ContextBag` on `ErrorContext`.

**Comment by andreasohlund** [comment]
Yea, thinking an extension would be a good way to expose it.

Then lets scope put the ErrorContext and add a core issue to get a
contextbag there in the future?

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers the [error context take a `TransportTransaction` as a parameter](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/ErrorContext.cs#L14) and is therefor extensible.

https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/TransportTransaction.cs

Would it be ok to propagate the `BasicDeliverEventArgs` there or would that be a hack? 

In short: Should I raise a core issue to consider adding a `ContextBag` to the `ErrorContext`?

**Comment by timbussmann** [comment]
> Would it be ok to propagate the `BasicDeliverEventArgs` there or would that be a hack?

Feels very much like a hack to me? +1 for raising a core issue.

**Comment by bording** [comment]
Yeah, I'd definitely consider that a hack and wouldn't want to do it.

**Comment by andreasohlund** [comment]
Ok, will raise a core issue

**Comment by andreasohlund** [comment]
Issue raised https://github.com/Particular/NServiceBus/issues/5333
